### PR TITLE
Centroids: transform orientations, esp. base grid

### PR DESCRIPTION
**Pull request by tovogt**
`Centroids.check` raises a `ValueError` when the raster's affine transform is not oriented up-down in y-direction. Before this PR, it would complain that the "upper left corner" has to be contained in the transform. With this PR the output message is improved to be more clear and the x-orientation is checked, as well.

Until we find the reason for this check (see #100), this PR proposes to modify`Centroids.get_closest_point` to work independent from transform orientation (following a suggestion by @ThomasRoosli). Furthermore, the `Centroids.from_base_grid` is changed to produce a `Centroids` with negative y-orientation. Finally, this PR implements unit tests for these changes. (Plus some additional code cleanup along the way.)

Edit: The PR has been changed to leave basic behavior as it used to be. It now mainly fixes the problems with `get_closest_point` and `from_base_grid`.

### CODE REVIEW

**Comment by chahank** [comment]
I am not sure to really understand the issue. Is there a good reason to allow not having the top left corner in?

File: climada/hazard/centroids/centr.py [hunk 1]
```diff
@@ -140,9 +142,9 @@ def check(self):
                 LOGGER.error('Missing meta information: width, height,'
                              'crs or transform')
                 raise ValueError
-            if self.meta['transform'][4] > 0:
-                LOGGER.error('Meta does not contain upper left corner data.')
-                raise ValueError
+            if self.meta['transform'][4] > 0 or self.meta['transform'][0] < 0:
+                LOGGER.warning('Meta does not contain upper left corner data.'
+                               'Some functionality might not work as expected.')
```
Comment (chahank) [inline climada/hazard/centroids/centr.py:147]: This warning is way too fuzzy and will most probably lead to issues in the future.

File: climada/hazard/centroids/centr.py [hunk 2]
```diff
@@ -22,20 +22,19 @@
...
+
+import geopandas as gpd
 import h5py
+import numpy as np
 import pandas as pd
-from rasterio import Affine
+import rasterio
 from rasterio.warp import Resampling
-import geopandas as gpd
+import scipy.sparse as sp
```
Comment (chahank) [inline climada/hazard/centroids/centr.py:32]: Probably better to keep is as `from scipy import sparse` as this seems to me the prevalent way in the current code base.

File: climada/hazard/centroids/centr.py (hunk 1 above)
Comment (tovogt) [reply climada/hazard/centroids/centr.py:147]: So do you suggest to raise an error or do you suggest do modify the output/explanation?

File: climada/hazard/centroids/centr.py [hunk 3]
```diff
@@ -140,9 +142,9 @@ def check(self):
                 LOGGER.error('Missing meta information: width, height,'
                              'crs or transform')
                 raise ValueError
-            if self.meta['transform'][4] > 0:
-                LOGGER.error('Meta does not contain upper left corner data.')
-                raise ValueError
+            if self.meta['transform'][4] > 0 or self.meta['transform'][0] < 0:
```
Comment (chahank) [inline climada/hazard/centroids/centr.py:145]: What are `self.meta['transform'][4]` and `self.meta['transform'][0]` ? Maybe defining variables for these makes is understandable.

File: climada/hazard/centroids/centr.py (hunk 2 above)
Comment (tovogt) [reply climada/hazard/centroids/centr.py]: Agreed and changed!

File: climada/hazard/centroids/centr.py (hunk 1 above)
Comment (chahank) [reply climada/hazard/centroids/centr.py:147]: The minimal solution is to provide a more detailed explanation, but is sub-optimal as the actual problem is not solved: why was there this constraint in the first place and do we need it?

Ideally, we want to know whether we need an error or whether the warning can be removed. I do not know the answer, however.

File: climada/hazard/centroids/centr.py (hunk 3 above)
Comment (tovogt) [reply climada/hazard/centroids/centr.py]: Agreed and changed!

File: climada/hazard/centroids/centr.py (hunk 1 above)
Comment (tovogt) [reply climada/hazard/centroids/centr.py:151]: Okay, then let's wait for answers in #100 before making decisions about this PR.

**Comment by tovogt** [comment]
Okay, @chahank: I modified the PR to be more along the lines of what @ThomasRoosli proposed in #100. Namely, we leave the basic behavior as is, improve the output messages a bit, add more unit tests, and (most importantly) make sure that the base grid passes the `check()`.

**Comment by chahank** [comment]
Cool thanks! I feel it is safer this way. It might be that we have to wait for the return of @gabrielaznar for a complete picture though.

File: climada/hazard/centroids/centr.py [hunk 4]
```diff
@@ -187,6 +194,21 @@ def from_base_grid(land=False, res_as=360, base_file=None):
...
+                centroids.meta['transform'] = rasterio.Affine(xres, xshear, xoff, yshear,
+                                                              -yres, yoff + (shape[0] - 1) * yres)
+                # flip y-axis in data arrays
+                for name in ["region_id", "dist_coast"]:
+                    if not hasattr(centroids, name):
+                        continue
+                    data = getattr(centroids, name)
+                    if data.size == 0:
+                        continue
+                    setattr(centroids, name, np.flipud(data.reshape(shape)).reshape(-1))
```
Comment (ThomasRoosli) [inline climada/hazard/centroids/centr.py:211]: Would it make sense to use centroids.check() before returning centroids? That would warn the user, if they assign a unknown file using the parameter base_file, and the resulting centroids are somehow not in order.

File: climada/hazard/centroids/test/test_centr.py [hunk 5]
```diff
@@ -70,21 +70,25 @@ def test_centroid_pass(self):
...
             extent=(-39, -34.7, -55.5, -53.6)  # south georgia island
         ).size
-
         self.assertEqual(count_sgi, 296)
 
+        # test negative latitudinal orientation by testing that northern hemisphere (Russia)
+        # is listed before southern hemisphere (South Africa)
+        russia_max_idx = (centroids.region_id == 643).nonzero()[0].max()
+        safrica_min_idx = (centroids.region_id == 710).nonzero()[0].min()
+        self.assertTrue(russia_max_idx < safrica_min_idx)
```
Comment (ThomasRoosli) [inline climada/hazard/centroids/test/test_centr.py:90]: Smart way of testing the (newly) required behaviour, together with the negative/zero diff in line 75 should work fine.

File: climada/hazard/centroids/test/test_vec_ras.py [hunk 6]
```diff
@@ -457,15 +460,36 @@ def test_size_pass(self):
...
+            ])
+            test_data[:,[1,3]] *= y_sign
+            for x_in, y_in, x_out, y_out, idx_out in test_data:
+                x, y, idx = centr_ras.get_closest_point(x_in, y_in)
+                self.assertEqual(x, x_out)
+                self.assertEqual(y, y_out)
+                self.assertEqual(idx, idx_out)
+                self.assertEqual(centr_ras.lon[idx], x)
+                self.assertEqual(centr_ras.lat[idx], y)
+
```
Comment (ThomasRoosli) [inline climada/hazard/centroids/test/test_vec_ras.py:486]: Good idea, to test this functionality also for the implemented changes

**Review by ThomasRoosli — APPROVED** [review]
First of all, sorry for the delayed review. Thanks for making the changes. I think this way we are playing it safe, and at the sametime do not have to change the binary basefile. Thanks for implementing it like that. I approve, but probably the develop branch has to be merged in first.

File: climada/hazard/centroids/centr.py (hunk 4 above)
Comment (tovogt) [reply climada/hazard/centroids/centr.py:211]: Yes, good idea! I added a check.

**Comment by tovogt** [comment]
Okay, thanks! This seems to be ready to go. The merge will be done after our meeting on Wednesday.

**Pull request merged**